cuda support fails when passed an object lib

blt add library with one of the sources like the following:

` $<TARGET_OBJECTS:blah>`

fails, in this macro:
```
macro(blt_split_source_list_by_language)

    set(options)
    set(singleValueArgs C_LIST Fortran_LIST)
    set(multiValueArgs SOURCES)

    # Parse the arguments
    cmake_parse_arguments(arg "${options}" "${singleValueArgs}"
                            "${multiValueArgs}" ${ARGN} )

    # Check arguments
    if ( NOT DEFINED arg_SOURCES )
        message( FATAL_ERROR "Must provide a SOURCES argument to the 'blt_split_source_list_by_language' macro" )
    endif()

    # Generate source lists based on language
    foreach(_file ${arg_SOURCES})
        get_filename_component(_ext ${_file} EXT)
        string(TOLOWER ${_ext} _ext_lower)

```

This returns empty:
` get_filename_component(_ext ${_file} EXT)`

Then, this fails:

`        string(TOLOWER ${_ext} _ext_lower)`


```
CMake Error at blt/cmake/BLTPrivateMacros.cmake:273 (string):
  string no output variable specified
Call Stack (most recent call first):
  blt/cmake/BLTPrivateMacros.cmake:226 (blt_split_source_list_by_language)
  blt/cmake/BLTMacros.cmake:456 (blt_setup_cuda_source_properties)
  ascent/CMakeLists.txt:286 (blt_add_library)

```
(Line numbers might be off, using older version of blt)

--- a/RELEASE-NOTES.md
+++ b/RELEASE-NOTES.md
@@ -30,6 +30,10 @@ The project release numbers follow [Semantic Versioning](http://semver.org/spec/
 - Improved how all BLT MPI information is being merged together and reported to users.
 - Increased number of ranks in `blt_mpi_smoke` test to catch regression.
 
+### Fixed
+- Do not attempt to split sources on generator expression
+
+
 ## 0.2.0 - Release date 2019-02-15
 
 ### Added

--- a/RELEASE-NOTES.md
+++ b/RELEASE-NOTES.md
@@ -29,9 +29,14 @@ The project release numbers follow [Semantic Versioning](http://semver.org/spec/
   generator expressions so thats a plus.
 - Improved how all BLT MPI information is being merged together and reported to users.
 - Increased number of ranks in `blt_mpi_smoke` test to catch regression.
+- blt_split_source_list_by_language now supports non-BLT object libraries and errors
+  out when any other generator expression is given in source list.  This is to avoid
+  very bad side effects of not being able to set source properties on anything
+  inside the generator expression.  This is because BLT cannot evaluate them.
 
 ### Fixed
-- Do not attempt to split sources on generator expression
+- Error out with better message when empty file extensions is hit in
+  blt_split_source_list_by_language.
 
 
 ## 0.2.0 - Release date 2019-02-15

--- a/cmake/BLTPrivateMacros.cmake
+++ b/cmake/BLTPrivateMacros.cmake
@@ -463,6 +463,12 @@ macro(blt_split_source_list_by_language)
 
     # Generate source lists based on language
     foreach(_file ${arg_SOURCES})
+        # Assume the user knows what they are doing if using a
+        # generator expressions because BLT can't evaluate them
+        if(${_file } MATCHES "^$<")
+            continue()
+        endif()
+
         get_filename_component(_ext ${_file} EXT)
         string(TOLOWER ${_ext} _ext_lower)
 

--- a/cmake/BLTPrivateMacros.cmake
+++ b/cmake/BLTPrivateMacros.cmake
@@ -465,7 +465,7 @@ macro(blt_split_source_list_by_language)
     foreach(_file ${arg_SOURCES})
         # Assume the user knows what they are doing if using a
         # generator expressions because BLT can't evaluate them
-        if(${_file } MATCHES "^$<")
+        if(${_file} MATCHES "^$<")
             continue()
         endif()
 

--- a/cmake/BLTPrivateMacros.cmake
+++ b/cmake/BLTPrivateMacros.cmake
@@ -445,6 +445,11 @@ endmacro(blt_add_hip_executable)
 ## blt_split_source_list_by_language( SOURCES <sources>
 ##                                    C_LIST <list name>
 ##                                    Fortran_LIST <list name>)
+##
+## Filters source list by file extension into C/C++ and Fortran source lists
+## based on BLT_C_FILE_EXTS and BLT_Fortran_FILE_EXTS (global BLT variables).
+## Files with no extension or generator expressions that are not object libraries
+## (of the form "$<TARGET_OBJECTS:nameofobjectlibrary>") will throw fatal errors.
 ##------------------------------------------------------------------------------
 macro(blt_split_source_list_by_language)
 
@@ -463,13 +468,19 @@ macro(blt_split_source_list_by_language)
 
     # Generate source lists based on language
     foreach(_file ${arg_SOURCES})
-        # Assume the user knows what they are doing if using a
-        # generator expressions because BLT can't evaluate them
-        if(${_file} MATCHES "^$<")
+        # Allow CMake object libraries but disallow generator expressions
+        # in source lists due to this causing all sorts of bad side effects
+        if("${_file}" MATCHES "^\\$<TARGET_OBJECTS:")
             continue()
+        elseif("${_file}" MATCHES "^\\$<")
+            message(FATAL_ERROR "blt_split_source_list_by_language macro does not support generator expressions because CMake does not provide a way to evaluate them. Given generator expression: ${_file}")
         endif()
 
         get_filename_component(_ext ${_file} EXT)
+        if("${_ext}" STREQUAL "")
+            message(FATAL_ERROR "blt_split_source_list_by_language given source file with no extension: ${_file}")
+        endif()
+
         string(TOLOWER ${_ext} _ext_lower)
 
         if(${_ext_lower} IN_LIST BLT_C_FILE_EXTS)

